rfd/0043-kubeaccess-multiparty.md

```diff
@@ -0,0 +1,387 @@
+
+It will work by adding a multiplexing layer inside the forwarder that similar to the current session recording
+functionality, but instead this multiplexes outputs to the session initiator and all observers
+but only streams back input from the initiator.
+
+#### Session observers
+
+A core feature we need to support is required observers. This will allow cluster administrators to configure
+policies that require certain Teleport users of a certain role to be actively monitoring the session. Each of these
+observers will also have the power to lock input/output for the session controller or instantly end it.
+
+This feature is useful in security critical environments where multiple people need to witness every action
+in the event of severe error or malice and have the ability to halt any erroneous or malicious action.
+
+#### Session states
+
+By default, a `tsh kube exec` and `kubectl attach` request will go through as usual if no policies are defined. If a policy like the one above is defined the session will be put in a pending state
+until the required viewers have joined.
+
+Sessions can have 3 possible states:
+
+- `PENDING`\
+  When a session is in a `PENDING` state, the connection to the pod from the proxy has not yet started
+  and all users are shown a default message informing them that the session is pending, current participants
+  and who else is required for the session to start.
+- `RUNNING`\
+A `RUNNING` session behaves like a normal multiparty session. `stdout`, `stdin` and `stdout` are mapped as usual
+  and the pod can be interacted with.
+- `TERMINATED`\
+  A session becomes `TERMINATED` once the shell spawned inside the pod quits or is forcefully terminated by one of the session participants.
+
+All sessions begin in the `PENDING` state and can change states based on the following transitions:
+
+##### Transition 1: `PENDING -> RUNNING`
+
+When the requirements for present viewers laid out in the role policy are fulfilled,
+the session transitions to a `RUNNING` state. This involves initiating the connection to the pod
+and setting up the shell. Finally, all clients are multiplexed
+onto the correct streams as described previously.
+
+Only the session initiator is able to make input, observers are not connected to the input stream
+and may only view stdout/stderr and terminate the session.
+
+##### Transition 2: `RUNNING -> TERMINATED`
+
+When the shell process created on the pod is terminated, the session transitions to a `TERMINATED` state and all clients
+are disconnected as per standard `kubectl` behaviour.
+
+##### Transition 3: `RUNNING -> TERMINATED`
+
+Session observers that are present may at any point decide to forcefully terminate the session.
+This will instantly disconnect input and output streams to prevent further communication. Once this is done
+the Kubernetes proxy will send a termination request to the pod session to request it be stopped.
```

Comment:
Just trying to understand the constraints is all. Killing subprocesses probably isn't even the right call, since some processes might leave things in an indeterminate state if killed before completion.